### ISSUE TITLE
Add an option to use the repository from the package.json

### ISSUE THREAD

**Issue by hitsthings** [webpro/release-it#42]
That would be the canonical repo, so would make sense (at least in my case) to use it as "src.pushRepo".

**Comment by webpro** [comment]
The default (for `git push`) is whatever is configured to be the push origin in the local repo (see `git remote -v`). Also, I think by default the `repository.url` in `package.json` should match this (not the other way around).

Apparently, you have a different setup, so you can override `src.pushRepo` any way you like.

Please let me know if I'm misunderstanding you. Feel free to re-open, but I'm closing it for now.

**Comment by hitsthings** [comment]
I normally have that setup too, and most people will. But the git configuration is defined by the local dev and the codebase itself has no input into how the local setup looks. Conversely, the package.json is defined by the codebase and can point to the known canonical repo no matter what the local setup looks like. If I'm publishing to NPM, then I want the tags in the canonical repo.


Take this situation.

1. git clone (origin URL) repo
2. cd repo
3. (Do some work but don't finish)
4. (Quick! There is a production bug to fix! Let's create a second working directory to deal with it)
5. cd ..
6. git clone ./repo repo2 # local clone is faster
7. cd repo2
8. (Fix bug quick!)
9. release-it patch

Now I've pushed the tag to ./repo instead of (origin URL) which is incorrect. If release-it used the package.json value I would have been fine.

Does this make sense?

You also mentioned that I can set src.pushRepo to whatever I want, but hardcoding the value twice in the codebase is not DRY and could cause issues if we move the repo later. What I've done so far is set up a script that reads the value from package.json and then programmatically calls release-it with a generated config.

I raised this improvement so it could be easier for others in the future. But if you see things differently that's cool. Cheers!

**Comment by webpro** [comment]
Regarding the extra local clone. In the situation you describe, I would *stash* my work, go to the branch, fix, release, change back to branch I was working in, apply stash.

And, imho the need to change this in two locations (if at all? YAGNI?) doesn't weigh up against creating that script. DRY is great, but not at all costs.

**Comment by hitsthings** [comment]
Thanks Lars, I think I just realized that `release-it` is not for me. Not because you're not implementing this (like I said, I'd already solved it), but because the value I was getting out of using it was already minimal, and it sounds like you're generally not interested in supporting more flexibility in workflows.

To address your two points (just for the sake of offering another perspective - I no longer have a vested interested in this):

> Regarding the extra local clone. In the situation you describe, I would stash my work, go to the branch, fix, release, change back to branch I was working in, apply stash.

Yeah that is sometimes better. But it completely ignores the cost of changing branches. On large projects, an initial build after checking out new code can take minutes. Having a second working directory available means you don't incur that cost when switching back to the feature you were working on. It's a valid workflow. The only reason it wouldn't be a valid workflow is that Git now has a feature to share a repo between two working directories. But most people don't know that and will still do the local clone - this is a footgun for them.

Of course this isn't the only case right? I could imagine someone has their own remote fork of the canonical repo. If they do a release locally, tags will be pushed to their fork and not the canonical. I'm sure there are other real world cases that would benefit from a codebase-controlled repo URL.

> And, imho the need to change this in two locations (if at all? YAGNI?) doesn't weigh up against creating that script. DRY is great, but not at all costs.

Repos change location all the time - when people switch hosting providers, or when, for example, Node was forked into IO.js, or again when the fork was merged. Offering one place to fix that would remove a potential bug for them.

It's also a bit of hyperbole to call this script "all costs" - it's really trivial to write. Here is one implementation (untested):

```js
if (config.src.pushRepo && config.src.pushRepo.usePackage) {
    config.src.pushRepo = JSON.parse(fs.readFileSync(path.join(process.cwd(), 'package.json'), 'utf8').repository.url;
}
```

It would support config like:

```
src : {
    pushRepo: { usePackage: true }
}
```

Anyway, best of luck!
